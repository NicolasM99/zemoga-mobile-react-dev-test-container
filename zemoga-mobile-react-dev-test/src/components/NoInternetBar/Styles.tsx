import { StyleSheet } from 'react-native';
export const noInternetBarStyles = StyleSheet.create({
  container: {
    height: 100,
    backgroundColor: 'red',
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingBottom: 8
  },
  message: {
    color: 'white',
    textAlign: 'center',
    margin: 0
  }
});
