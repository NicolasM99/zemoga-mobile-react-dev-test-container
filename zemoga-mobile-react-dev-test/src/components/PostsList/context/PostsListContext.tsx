import React, { FC, useState, createContext, useContext, useEffect } from 'react';

import { generalAction } from 'src/redux/actions';
import { GET_POSTS } from 'src/redux/actionTypes';
import { useRedux } from 'src/redux/hooks/useRedux';
import { POSTS_API } from 'src/redux/paths';

import { PostsListProviderProps, PostsListContextType, IPost } from '../@types/postListContext';

const PostsListContext = createContext<PostsListContextType | null>(null);

const PostsListProvider: FC<PostsListProviderProps> = ({ children }: PostsListProviderProps) => {
  const [posts, setPosts] = useState<IPost[] | []>([]);
  const [postsToDelete, setPostsToDelete] = useState<IPost[] | []>([]);
  const [isDeletingItems, setIsDeletingItems] = useState(false);
  const [dispatch, { posts: postsStore }, status] = useRedux(['posts']);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  useEffect(() => {
    if (postsStore) setPosts([...postsStore]);
  }, [postsStore]);

  useEffect(() => {
    if (status) setIsLoading(false);
  }, [status]);

  useEffect(() => {
    if (isLoading) setIsLoading(false);
  }, [isLoading]);

  useEffect(() => {
    dispatch(generalAction({ actionType: GET_POSTS, api: POSTS_API }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const valueObj = {
    posts,
    setPosts,
    isDeletingItems,
    setIsDeletingItems,
    postsToDelete,
    setPostsToDelete,
    isLoading,
    setIsLoading
  };
  return <PostsListContext.Provider value={valueObj}>{children}</PostsListContext.Provider>;
};

const usePostsListContext = () => useContext(PostsListContext);

export { PostsListContext, PostsListProvider, usePostsListContext };
