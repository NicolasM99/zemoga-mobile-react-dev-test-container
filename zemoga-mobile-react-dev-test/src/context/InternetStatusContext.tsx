import React, { createContext, FC, useContext, useEffect, useState } from 'react';

import NetInfo from '@react-native-community/netinfo';

import { IInternetStatusContext, InternetStatusContextType } from './@types/internetStatusContext';

const InternetStatusContext = createContext<InternetStatusContextType | null>(null);

const InternetStatusProvider: FC<IInternetStatusContext> = ({
  children
}: IInternetStatusContext) => {
  const [internetStatus, setInternetStatus] = useState<boolean | null>(true);
  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener((state) => {
      setInternetStatus(state.isConnected);
    });
    return () => {
      unsubscribe();
    };
  }, []);

  const valueObj = {
    internetStatus
  };
  return (
    <InternetStatusContext.Provider value={valueObj}>{children}</InternetStatusContext.Provider>
  );
};

const useInternetStatusContext = () => useContext(InternetStatusContext);

export { InternetStatusContext, InternetStatusProvider, useInternetStatusContext };
