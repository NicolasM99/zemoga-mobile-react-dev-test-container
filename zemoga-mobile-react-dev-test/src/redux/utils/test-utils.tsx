import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { render } from '@testing-library/react-native';
import type { RenderOptions } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import { INITIAL_STATE } from '../reducers';
import { persistedReducer, store as reduxStore } from '../store';
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<typeof persistedReducer>;
  store?: typeof reduxStore;
}

export const renderWithStoreProviders = (
  ui: React.ReactElement,
  {
    preloadedState = INITIAL_STATE,
    store = configureStore({ reducer: persistedReducer, preloadedState }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<object>): JSX.Element => (
    <Provider store={store}>{children}</Provider>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
