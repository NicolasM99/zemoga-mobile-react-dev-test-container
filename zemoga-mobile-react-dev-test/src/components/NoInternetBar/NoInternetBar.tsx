import React, { FC } from 'react';

import { View } from 'react-native';

import { InternetStatusContextType } from 'src/context/@types/internetStatusContext';
import { useInternetStatusContext } from 'src/context/InternetStatusContext';

import { noInternetBarStyles } from './Styles';
import Header from '../Header/Header';

const NoInternetBar: FC = () => {
  const { internetStatus } = useInternetStatusContext() as InternetStatusContextType;
  if (!internetStatus)
    return (
      <View style={noInternetBarStyles.container}>
        <Header variant="h3" bold={false} style={noInternetBarStyles.message}>
          No internet access. You're seeing either old posts or no author/comments information and
          you can't perform update or delete actions.
        </Header>
      </View>
    );
  return <></>;
};

export default NoInternetBar;
