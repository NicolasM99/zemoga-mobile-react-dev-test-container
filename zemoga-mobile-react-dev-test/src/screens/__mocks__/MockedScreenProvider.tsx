import React from 'react';

import { NavigationContainer } from '@react-navigation/native';

import { PostsListProvider } from 'src/components/PostsList/context/PostsListContext';
import { InternetStatusProvider } from 'src/context/InternetStatusContext';

interface MockedScreenProviderProps {
  children: React.ReactNode;
}

const MockedScreenProvider: React.FC<MockedScreenProviderProps> = ({
  children
}: MockedScreenProviderProps) => (
  <NavigationContainer>
    <InternetStatusProvider>
      <PostsListProvider>{children}</PostsListProvider>
    </InternetStatusProvider>
  </NavigationContainer>
);

export default MockedScreenProvider;
