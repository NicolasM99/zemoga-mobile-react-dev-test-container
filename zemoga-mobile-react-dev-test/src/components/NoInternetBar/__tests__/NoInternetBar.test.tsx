import { screen } from '@testing-library/react-native';

import * as PostsListContext from 'src/components/PostsList/context/PostsListContext';
import * as InternetStatusContext from 'src/context/InternetStatusContext';
import { renderWithStoreProviders } from 'src/redux/utils/test-utils';
import MockedScreenProvider from 'src/screens/__mocks__/MockedScreenProvider';

import NoInternetBar from '../NoInternetBar';

describe('<NoInternetBar /> unit tests', () => {
  jest.mock('react-redux', () =>
    jest.fn(() => ({
      useDispatch: () => true
    }))
  );
  jest.spyOn(InternetStatusContext, 'useInternetStatusContext').mockReturnValue({
    internetStatus: false
  });
  jest.spyOn(PostsListContext, 'usePostsListContext').mockReturnValue({
    isLoading: false,
    posts: [],
    setPosts: () => {},
    isDeletingItems: false,
    setIsDeletingItems: () => {},
    postsToDelete: [],
    setPostsToDelete: () => {},
    setIsLoading: () => {}
  });
  it('Should mount correctly the component', () => {
    renderWithStoreProviders(
      <MockedScreenProvider>
        <NoInternetBar />
      </MockedScreenProvider>
    );
    expect(screen).toMatchSnapshot();
  });

  it('Should return empty object if internetStatus is true', () => {
    jest.spyOn(InternetStatusContext, 'useInternetStatusContext').mockReturnValue({
      internetStatus: true
    });
    renderWithStoreProviders(
      <MockedScreenProvider>
        <NoInternetBar />
      </MockedScreenProvider>
    );
    expect(screen).toMatchObject({});
  });
});
