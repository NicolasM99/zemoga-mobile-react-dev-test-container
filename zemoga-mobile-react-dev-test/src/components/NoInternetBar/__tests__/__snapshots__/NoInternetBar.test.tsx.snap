// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NoInternetBar /> unit tests Should mount correctly the component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "red",
      "justifyContent": "flex-end",
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 64,
    }
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 16,
        "fontWeight": "normal",
        "marginBottom": 8,
        "textAlign": "center",
      }
    }
    testID="header"
  >
    No internet access. You're seeing either old posts or no author/comments information and you can't perform update or delete actions.
  </Text>
</View>
`;
